spec/ParseFile.spec.js

```diff
@@ -860,4 +860,162 @@ describe('Parse.File testing', () => {
       });
     });
   });
+
+  describe('disable file upload', () => {
+    it('can reject file upload with unspecified', async () => {
+      await reconfigureServer({
+        fileUpload: {},
+      });
+      try {
+        const file = new Parse.File('hello.txt', data, 'text/plain');
+        await file.save();
+        fail('should not have been able to save file.');
+      } catch (e) {
+        expect(e.code).toBe(130);
+        expect(e.message).toBe('File upload by public is not enabled.');
+      }
+    });
+
+    it('disable all file upload', async () => {
+      await reconfigureServer({
+        fileUpload: {
+          enableForPublic: false,
+          enableForAnonymousUser: false,
+          enableForAuthenticatedUser: false,
+        },
```

Comment:
DX: Change the interface to get something more close to the current CLPs interface

```js
fileLevelPermissions: {
  upload: {
      requiresAuthentication: true
      public: false
      anonymous: false
  }
}
```

It will also allow us to support easly without breaking the interface the get permissions
When access to the file do is not direct it could be interesting to support the `get` permissions

```js
fileLevelPermissions: {
  upload: {
      requiresAuthentication: true
      public: false
      anonymous: false
  },
  get: { 
    public: false
  }
}
```

Reply:
What are your thoughts on this @mtrezza?

The “get” file permission will probably require some temporary token to differentiate what user is requesting the file. [(See this comment)](https://github.com/parse-community/parse-server/issues/7001#issuecomment-743973199)

Reply:
>  Change the interface to get something closer to the current CLPs interface

Without anticipating the [config structure discussion](https://github.com/parse-community/parse-server/issues/7069), I am not in favor of less specific environment keys such as `enable`, `public`, `anonymous`, etc. While this makes sense in an OO coding environment, it seems too unspecific for environment variables and potentially create extensibility issues for the same parameter group.

I suggest to stick to config parameter that goes max. 1 level deep, for simplicity, consistency with the current config structure and because we did not evaluate the necessary changes for deeper nested config parameters and potential effects on existing deployments where these values are set via environment variables rather than in code. For example, I came across confirmed limitations in AWS EB environments when setting environment variables to stringified object values.

> When access to the file is not direct (and managed by Parse Server) it could be interesting to support the get permissions

I suggest to keep this PR to its original scope while keeping in mind a potential future PR for file download restriction. It has proved in the past to be beneficial to not extend PR scopes on the fly for efficiency and oversight. I suggest to address this in a potential follow-up PR.

Reply:
Yes the `get` suggestion was a `feature request`.

In other hand we have introduced some security concepts like ACL/CLP; i think we should keep File Security interface close to the CLP interface, then developers need only to learn one interface.

About the AWS issue, i think we should not take it as a constraint but as a way of reflexion. Meteor team got the same issue with the `METEOR_SETTINGS` env var.

We can additionally support this kind of env structure for developers that encounter some issue with JSON config.
Internally parse server could convert `FILE_LEVEL_PERMISSIONS_UPLOAD_PUBLIC=0` to `{ fileLevelPermissions: {upload: {public: false } }`  parsing dynamically `FILE_LEVEL_PERMISSIONS_UPLOAD_PUBLIC` to a JSON structure is pretty easy :)

Reply:
> We can additionally support...

Let's keep this PR in scope please.

Reply:
Please @mtrezza , do not resolve conversation until it's finished and we get some other feedbacks on this from other contributors :)

About the AWS issue it's just a proposal on how to fix it later , in another PR.
I just think here we need to take a step back to get a global view. Since we saw that any change into ParseOptions will lead to a breaking change on deployments.

Based on our context, currently we have CLP interface for `Parse.User`, `Parse.Object`, `Parse.Role`, why getting a different interface for `Parse.File`

Also does the ParseOptions is the best way to set this kind of permissions ? Since a change need a full re deploy, may be it's better way to implement this feature with an API route and store these options into `_SCHEMA` collection. Then parameters could be managed easily with another PR on the Parse Dashboard ? And also add a method to `Parse.File` on the SDK JS.

From my point of view the best interface will be something like

```
// Need master key
Parse.File.setCLP({ create: { requiresAuthentication: true, "*": false }})
```

Reply:
Additional note using the standard CLP interface allow to use the standard CLP validator and a much more flexible permission like

```
// Need master key
Parse.File.setCLP({ create: { requiresAuthentication: false, "*": false, 'role:Admin': true }})
```

Reply:
I appreciate your input and these are good ideas worth considering.

However, there good reasons to have this discussion as a separate issue:
- Improving Parse Server security features is a strategic priority. Your requested changes deviate substantially from this PR and requires additional, prolonged effort to achieve the same outcome at a later point in time. This PR is close to ready for merge.
- Parse.File is not a class. Using the CLP interface and adding this to schema may collide with other features in the future, such as [auto-deletion of files](https://github.com/parse-community/parse-server/issues/6780) where we may need an actual `_File` class. It would be the only entry in the `_SCHEMA` that is not an actual class which is a deviation from current concepts and requires deliberation.

I invite you to have this discussion in a separate issue, would you be open to that?

Reply:
> Your requested changes deviate substantially from this PR and require additional, prolonged effort to achieve the
> same outcome at a later point in time. 

Yes it need more work, but Parse server is currently fine without this feature and we have some workarounds (like triggers) to easily fix the file permission issue while waiting the new permission level interface. I'm not in favor into "rushing" features but into focusing to implement features in a durable way (like Auth Adapter improved spec). This will reduce and avoid future rework, and save developers from using an interface that may soon be obsolete.  As we are a volunteer Open Source team, we have the time to work properly on features, we have no deadlines. I think we should do it properly with full support on the JS SDK.

And yes an issue is better to define the spec and avoid wasting time on a wrong implementation, i will be happy to work on this proposal 😃